src/dev/run_check_lockfile_symlinks.js

```diff
@@ -0,0 +1,182 @@
+  await checkOnlyLockfileAtProjectRoot(filtered);
+  await checkSuperfluousSymlinks(log, filtered);
+  await checkMissingSymlinks(log, filtered);
+}
+
+async function checkOnlyLockfileAtProjectRoot(files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), LOCKFILE_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const stats = lstatSync(path);
+      if (!stats.isSymbolicLink() && parent !== '.') {
+        errorPaths.push(path);
+      }
+    });
+
+  if (errorPaths.length) {
+    throw createFailError(
+      `These directories MUST NOT have a 'yarn.lock' file:\n${listPaths(errorPaths)}`
+    );
+  }
+}
+
+async function checkSuperfluousSymlinks(log, files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), LOCKFILE_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const stats = lstatSync(path);
+      if (!stats.isSymbolicLink()) {
+        return;
+      }
+
+      const manifestPath = `${parent}/package.json`;
+      if (!existsSync(manifestPath)) {
+        log.warning(
+          `No manifest found at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+        );
+        errorPaths.push(path);
+        return;
+      }
+
+      try {
+        const manifest = readFileSync(manifestPath);
+        try {
+          const json = JSON.parse(manifest);
+          if (!json.dependencies || !Object.keys(json.dependencies).length) {
+            log.warning(
+              `Manifest at '${manifestPath}' has no dependencies, but found an adjacent 'yarn.lock' symlink.`
+            );
+            errorPaths.push(path);
+          }
+        } catch (err) {
+          log.warning(
+            `Could not parse JSON at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+          );
+          errorPaths.push(path);
+        }
+      } catch (err) {
+        log.warning(
+          `Could not read manifest at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+        );
+        errorPaths.push(path);
+      }
+    });
+
+  if (errorPaths.length) {
+    throw createFailError(
+      `These directories MUST NOT have a 'yarn.lock' symlink:\n${listPaths(errorPaths)}`
+    );
+  }
+}
+
+async function checkMissingSymlinks(log, files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), MANIFEST_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const lockfilePath = `${parent}/yarn.lock`;
+      if (existsSync(lockfilePath)) {
+        return;
+      }
+
+      try {
+        const manifest = readFileSync(path);
+        try {
+          const json = JSON.parse(manifest);
+          if (json.dependencies && Object.keys(json.dependencies).length) {
```

Comment:
Couldn't we have a scenario where there's `devDependencies`, but no `dependencies` and wouldn't that be ok?

Reply:
It would be OK, but since this PR is to support scanning tools, we generally don't care about dev dependencies for those.

Similar to my answer above, the rationale behind this check is that if a package has `dependencies`, it needs a lockfile symlink next to it. Otherwise, if it _doesn't_ have `dependencies`, it doesn't need a symlink here.